src/daal4py.cpp

```diff
@@ -244,75 +222,97 @@ extern PyObject * make_nda(daal::data_management::KeyValueDataCollectionPtr * di
         } else {
-            std::cerr << "Input array has wrong dimensionality (must be 2d).\n";
+            std::cerr << "Input array is neither well behaved and nor a structured array.\n";
         }
+    } else {
+        std::cerr << "Input array has wrong dimensionality (must be 2d).\n";
     }
 
-    if(!ptr) std::cerr << "Could not convert Python object to DAAL table.\n";
-
     return ptr;
 }
 
 daal::data_management::NumericTablePtr * make_nt(PyObject * nda)
 {
     if(nda && nda != Py_None) {
         daal::data_management::NumericTable * ptr = NULL;
+        if(is_array(nda)) { // we got a numpy array
+            PyArrayObject * ary = (PyArrayObject*)nda;
 
-        switch(array_type(nda)) {
-        case NPY_UINT32:
-            ptr = _make_nt<uint32_t, NPY_UINT32>(nda);
-            break;
-        case NPY_INT32:
-            ptr = _make_nt<int32_t, NPY_INT32>(nda);
-            break;
-        case NPY_UINT64:
-            ptr = _make_nt<uint64_t, NPY_UINT64>(nda);
-            break;
-        case NPY_INT64:
-            ptr = _make_nt<int64_t, NPY_INT64>(nda);
-            break;
-        case NPY_FLOAT32:
-            ptr = _make_nt<float, NPY_FLOAT32>(nda);
-            break;
-        default:
-            ptr = _make_nt<double, NPY_FLOAT64>(nda);
-        }
+            if(array_is_behaved(ary)) {
+#define MAKENT_(_T) ptr = _make_hnt<_T>(nda)
+                SET_NPY_FEATURE(PyArray_DESCR(ary)->type, MAKENT_, throw std::invalid_argument("Found unsupported array type"));
+#undef MAKENT_
+            } else {
+                ptr = _make_npynt(nda);
+            }
+
+            if(!ptr) std::cerr << "Could not convert Python object to DAAL table.\n";
+
+        } else if(PyList_Check(nda) && PyList_Size(nda) > 0) { // a list of arrays for SOA?
+            PyObject * first = PyList_GetItem(nda, 0);
+
+            if(is_array(first)) { // can handle only list of 1d arrays
+                auto N = PyList_Size(nda);
+                daal::data_management::SOANumericTable * soatbl = NULL;
 
+                for(auto i = 0; i < N; i++) {
+                    PyArrayObject * ary = (PyArrayObject*)PyList_GetItem(nda, i);
+                    if(i==0) soatbl = new daal::data_management::SOANumericTable(N, PyArray_DIMS(ary)[0]);
+                    if(PyArray_NDIM(ary) != 1) {
+                        std::cerr << "Found wrong dimensionality (" << PyArray_NDIM(ary) << ") of array in list when constructing SOA table (must be 1d)";
+                        delete soatbl;
+                        soatbl = NULL;
+                        break;
+                    }
+
+#define SETARRAY_(_T) {daal::services::SharedPtr< _T > _tmp(reinterpret_cast< _T * >(PyArray_DATA(ary)), NumpyDeleter(ary)); soatbl->setArray(_tmp, i);}
+                    SET_NPY_FEATURE(PyArray_DESCR(ary)->type, SETARRAY_, throw std::invalid_argument("Found unsupported array type"));
+#undef SETARRAY_
+
```

Comment:
@RuslanIsrafilov is it enough to create a SOANumericTable with size and then just set the column pointers? Or do we need something else?
Note: the columns pointers are shared-pointers. daal4py will hold a ref until the numpy array gets GC'ed/deleted.

Reply:
It is enough